src/test/java/com/epam/esm/service/certificate/GiftServiceImplTest.java

```diff
@@ -0,0 +1,294 @@
+package com.epam.esm.service.certificate;
+
+import com.epam.esm.entity.DateSortType;
+import com.epam.esm.entity.GiftCertificate;
+import com.epam.esm.entity.Tag;
+import com.epam.esm.entity.dto.GiftCertificateDto;
+import com.epam.esm.entity.dto.RequestParametersDto;
+import com.epam.esm.entity.dto.TagDto;
+import com.epam.esm.exception.EntityNotAddedException;
+import com.epam.esm.exception.EntityNotDeletedException;
+import com.epam.esm.exception.EntityNotFoundException;
+import com.epam.esm.exception.EntityNotUpdatedException;
+import com.epam.esm.repository.CriteriaSpecification;
+import com.epam.esm.repository.certificate.CertificateRepositoryImpl;
+import com.epam.esm.repository.specification.CertificatesByTagNameCriteriaSpecification;
+import com.epam.esm.repository.specification.SpecificationCreator;
+import com.epam.esm.service.mapper.certificate.CertificateMapper;
+import com.epam.esm.service.mapper.tag.TagMapper;
+import com.epam.esm.service.tag.TagServiceImpl;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.InjectMocks;
+import org.mockito.Mock;
+import org.mockito.Mockito;
+import org.mockito.junit.jupiter.MockitoExtension;
+
+import java.math.BigDecimal;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+
+@ExtendWith(MockitoExtension.class)
+class GiftServiceImplTest {
+  @InjectMocks
+  private GiftServiceImpl service;
+  @Mock
+  private CertificateRepositoryImpl repository;
+  @Mock
+  private CertificateMapper mapper;
+  @Mock
+  private TagMapper tagMapper;
+  @Mock
+  private TagServiceImpl tagService;
+  @Mock
+  private SpecificationCreator creator;
+
+  @Test
+  void addCertificate_ShouldReturnCertificate_True_test() {
+    String expectedName = "name1";
+    GiftCertificate certificate = new GiftCertificate();
+    certificate.setName(expectedName);
+    certificate.setDescription("description1");
+    certificate.setPrice(new BigDecimal("100.10"));
+    certificate.setDuration(10);
+    certificate.setGiftId(1L);
+
+    Tag tag = new Tag();
+    tag.setName("name1");
+
+    TagDto tagDto = new TagDto("name1");
+    tagDto.setId(1);
+    List<TagDto> dtos = Collections.singletonList(tagDto);
+
+    GiftCertificateDto certificateDto =
+        new GiftCertificateDto("name1", "description1", new BigDecimal("100.10"), 10, dtos);
+
+    Mockito.when(repository.add(certificate)).thenReturn(Optional.of(certificate));
+    Mockito.when(mapper.toEntity(certificateDto)).thenReturn(certificate);
+    Mockito.when(mapper.toDto(certificate)).thenReturn(certificateDto);
+    Mockito.when(tagService.addTagIfNotExist(Mockito.any(TagDto.class))).thenReturn(tagDto);
+    Mockito.when(tagMapper.toEntity(tagDto)).thenReturn(tag);
+
+    String actualName = service.add(certificateDto).getName();
+
+    Assertions.assertEquals(expectedName, actualName);
```

Comment:
I guess that we we have to use `verify` method to control that mock was called

Reply:
Fixed